sycl/doc/extensions/SubGroup/SYCL_INTEL_sub_group.asciidoc

```diff
@@ -68,7 +68,9 @@ Providing a generic group abstraction encapsulating the shared functionality of
 
 === Attributes
 
-The +[[intel::reqd_sub_group_size(n)]]+ attribute indicates that the kernel must be compiled and executed with a sub-group of size _n_.  The value of _n_ must be a compile-time integral constant expression.  The value of _n_ must be set to a sub-group size that is both supported by the device and compatible with all language features used by the kernel, or device compilation will fail.  The set of valid sub-group sizes can be queried as described below.
+The +[[intel::reqd_sub_group_size(S)]]+ attribute indicates that the kernel must be compiled and executed with a specific sub-group size.  The value of _S_ must be a compile-time integral constant expression or one of the following keywords: +auto+, +core+.  If the value of _S_ is an integer, the kernel should only be submitted to a device that supports that sub-group size (as reported by +info::device::sub_group_sizes+).  If the kernel is submitted to a device that does not support the requested sub-group size, or a device on which the requested sub-group size is incompatible with any language features used by the kernel, the implementation must throw a synchronous exception with the `errc::feature_not_supported` error code from the kernel invocation command.  If the value of _S_ is +auto+, the implementation is free to select any of the valid sub-group sizes associated with the device to which the kernel is submitted; the manner in which the sub-group size is selected is implementation-defined.  If the value of _S_ is +core+, the implementation will select the device's core sub-group size (as reported by the +info::device::core_sub_group_size+ query) for all kernels with this attribute.
+
+If no required sub-group size attribute appears on a kernel, the default behavior is as-if +[[intel::reqd_sub_group_size(auto)]]+ was specified.  This behavior may be overridden by an implementation (e.g. via compiler flags).
 
 In addition to device functions, the required sub-group size attribute may also be specified in the definition of a named functor object, as in the example below:
```

Comment:
The rest of `Attributes` section below always bothered me and having more values for `intel::reqd_sub_group_size` attribute concerns me even more:

> The +[[intel::reqd_sub_group_size(S)]]+ attribute indicates that the kernel must be compiled and executed with a specific sub-group size.
> In addition to device functions,

At first, we talk about _kernels_  and then we say "In addition to device functions", which implies that the attribute can also be applied to arbitrary device function and this particular usage is the least documented one. 

If I understand correctly, the main idea of this attribute is to control sub-group size of the particular _kernel_,  but of an arbitrary function: I would suggest that we remove this "In addition to device function" and if want to say something about arbitrary device functions, let's say it below in a direct manner, probably with examples.

> It is illegal for a kernel or function to call a function with a mismatched sub-group size requirement, and the compiler should produce an error in this case.  The +reqd_sub_group_size+ attribute is not propagated from a device function to callers of the function, and must be specified explicitly when a kernel is defined.

How these restrictions should work with `auto` and `core`? I mean things like can we call `core` function from an `auto` function? Can we call `core` function from a function with sub-group size `8`, for example?

Should compiler trace the whole call graph? I.e if we have `bar [[reqd_sub_group_size(8)]]` calling `foo` (no required sub-group size), which in turn calls `baz [[reqd_sub_group_size(4)]]` - should error be reported by the compiler? Or it is up to implementation to decide about quality of error checking?

Reply:
> I would suggest that we remove this "In addition to device function" and if want to say something about arbitrary device functions, let's say it below in a direct manner, probably with examples.

You're right.  The intent here was to allow `reqd_sub_group_size` to be specified on `SYCL_EXTERNAL` functions, but that's not spelled out explicitly here and I don't think it's implemented either.  @gmlueck, @rolandschulz , do you think it would be okay to change this to only allow `reqd_sub_group_size` on external functions (rather than arbitrary functions)?  That makes the intent clearer.

> How these restrictions should work with auto and core? I mean things like can we call core function from an auto function? Can we call core function from a function with sub-group size 8, for example?

Good question.  I can see an argument either way.  On the one hand, if `core` happens to be 8 on some device, linking makes sense; similarly, calling a function with a required size of 8 might encourage `auto` to select 8 (if it can).  On the other hand, whether `core` and `auto` are compatible with a specific size cannot be determined until the device is known.

It seems to me like the safest option would be to only allow exact matches, so a kernel with `auto` can only call an external function with `auto` and so on.  Do you agree?

> Should compiler trace the whole call graph? I.e if we have bar [[reqd_sub_group_size(8)]] calling foo (no required sub-group size), which in turn calls baz [[reqd_sub_group_size(4)]] - should error be reported by the compiler? Or it is up to implementation to decide about quality of error checking?

Changing the wording as above would simplify this.  The compiler would have to trace the whole call graph, but would only have to check for mismatched sub-group sizes when encountering an external function.  Does that make sense?

Reply:
> It seems to me like the safest option would be to only allow exact matches, so a kernel with auto can only call an external function with auto and so on. Do you agree?

I think that we should allow calling `auto` from non-`auto`, but shouldn't allow calling non-`auto` from `auto`. So, if user has a kernel, which requires sub-group size `8` or `core`, user is still able to call function with `auto` (either explicit or implicit) sub-group size - IMHO, it makes sense to not require from user to put the attribute on *every* function in a call tree - it is enough to just mark interfaces (like kernels and `SYCL_EXTERNAL` functions)

However, if user has `auto` function, it should be prohibited to call `core` or `8` functions from it, i.e. user is either only marks interfaces, or propagates this attribute down by call graph to any depth it wants to, but without any "gaps". By doing so, we still allowing users to put the attribute everywhere to improve code readability (i.e. you don't need to manually trace the call graph upwards to see which sub-group size is going to be used for this function) and also we could simplify the compiler as it would be able to only check caller-callee pair and wouldn't be required to trace the whole call graph.

What do you think?

> The compiler would have to trace the whole call graph, but would only have to check for mismatched sub-group sizes when encountering an external function. Does that make sense?

We might have `auto` function, which calls `8`, but itself is called from `4` and `16. This doesn't look correct and I think that compiler should report errors in this case.

One more question, consider the following situation:  kernel with `8` calls function `foo [auto]` and kernel with `4` calls the same `foo [auto]` - is such configuration considered to be a consistent? Is it important to document expected behavior? I mean SPIR-V spec says that such configuration is valid and `foo` must be cloned and compiled two times for different sub-group sizes (because sub-group size is a property of `EntryPoint`)

Reply:
The cleanest solution would be to allow `[[sycl::reqd_sub_group_size()]]` only on kernel declarations, which is how it is currently specified in SYCL 2020 provisional.  If the implementation supports `SYCL_EXTERNAL`, the responsibility would be on the implementation to propagate the sub-group size from the kernel to any `SYCL_EXTERNAL` functions it calls.  This would require the implementation to have some sort of phase the merges SPIR-v across translation units.  It seems like any implementation that supports `SYCL_EXTERNAL` must have some sort of phase like this, though, right?

If we think that our implementation cannot do this in the short term, we have two options:

1. Say that `[[sycl::reqd_sub_group_size()]]` is not supported for kernels that call `SYCL_EXTERNAL` functions.  The compiler should raise a diagnostic in this case.

2. Invent a syntax where `[[sycl::reqd_sub_group_size()]]` can be used to decorate `SYCL_EXTERNAL` functions, as we're describing above.  In this case, we should plan for this syntax to be optional in the future.  In the future, I think it would be better for the implementation to automatically propagate the sub-group information to `SYCL_EXTERNAL` functions.

If we decide on option 2, then we should come up with the simplest syntax that addresses the issue.  This is a temporary solution, so we don't need it to be too fancy.  I would be tempted to agree with @Pennycook that we should allow only an exact match between a kernel's sub-group size and a `SYCL_EXTERNAL` function's sub-group size.  However, what about the case where two kernels call the same `SYCL_EXTERNAL` function?  What if those two kernels had different `[[sycl::reqd_sub_group_size()]]` attributes?

One way to solve this would be to allow a list of sizes in `[[sycl::reqd_sub_group_size()]]` only when it is applied to a `SYCL_EXTERNAL` function.  This would tell the compiler to produce multiple versions of the function, and each kernel could use the appropriate version.

Reply:
> One more question, consider the following situation: kernel with 8 calls function foo [auto] and kernel with 4 calls the same foo [auto] - is such configuration considered to be a consistent? Is it important to document expected behavior? I mean SPIR-V spec says that such configuration is valid and foo must be cloned and compiled two times for different sub-group sizes (because sub-group size is a property of EntryPoint)

This is a great example of why I think we want to tie this to `SYCL_EXTERNAL` somehow.  If a compiler wants to detect that a function is called from multiple kernels and multi-version that function, it should do that.  But as soon as a function is marked external, how to multi-version becomes less obvious.  For SPIR-V it may be fine, but what about other implementations? How does it interact with [function pointers](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/SPIRV/SPV_INTEL_function_pointers.asciidoc)?  On the flip side, allowing `auto` on an external function might not make a lot of sense either -- the compiler can definitely choose a sub-group size to compile the function with, but the user can no longer reason about which kernels can call it.

> I think it would be better for the implementation to automatically propagate the sub-group information to SYCL_EXTERNAL functions.

I don't think this is realistic for all implementations.  If kernels and external functions are compiled straight to some ISA for a specific device, there's no opportunity to do this propagation.  If the kernel could prove which function(s) it was calling and which needed to be compiled for the device, there'd be no need to mark the function(s) with `SYCL_EXTERNAL`.

I'm a little scared to try and solve the problem for the generic case here: there are clearly a lot of open questions, and I don't want to set a precedent for all other extensions and attributes in DPC++ without more discussion.  That said -- @kbobrovs, am I right in thinking that ESIMD is currently checking that kernels and external functions have an exact match of `reqd_sub_group_size(1)`?

If so, the simplest thing may be to say:
- `[[sycl::reqd_sub_group_size()]]` is allowed on kernels and `SYCL_EXTERNAL` functions (but not normal device functions).
- `[[sycl::reqd_sub_group_size(auto)]]` is invalid for `SYCL_EXTERNAL` functions.
- If a kernel calls a `SYCL_EXTERNAL` function, the attributes applied to each must match exactly.

I could then add "How does sub-group size interact with `SYCL_EXTERNAL` functions?" to the list of unresolved questions for this extension, to be addressed later.  It seems to me like exact matching or multi-versioning could work, but we should consider more than just sub-group size when we make the decision.

Reply:
I'm OK with that for the short-term solution.

Reply:
Applied in 6482dcfd262071b8eddedb15cd6db4ae19bf8cdf.